{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2bc102eb-fd66-46a4-8e6f-c43b16c25e76",
   "metadata": {},
   "source": [
    "## Collecting data from gnomAD\n",
    "\n",
    "Variants data downloaded from [gnomAD](https://gnomad.broadinstitute.org/downloads#v4)\n",
    "\n",
    "Use command line for initial processing\n",
    "\n",
    "1. Filtered only PASS variants \n",
    "\n",
    "```zsh\n",
    "bcftools view -f 'PASS,.' gnomad.exomes.v4.0.sites.chr22.vcf.bgz > filtered_gnomad22.bgz\n",
    "```\n",
    "For next parcing we can use also **command line** or **python**\n",
    "\n",
    "2. Extract the necessary data\n",
    "\n",
    "for this step we need **bcftools** utility\n",
    "\n",
    "```zsh\n",
    "bcftools query -f '%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%AC\\t%AC_afr\\t%AC_amr\\t%AC_nfe\n",
    "\\t%AC_asj\\t%AC_sas\\t%AC_eas\\t%AC_mid\\t%AC_fin\\t%AN\\t%AN_afr\\t%AN_amr\\t%AN_nfe\\t%AN_asj\n",
    "\\t%AN_sas\\t%AN_eas\\t%AN_mid\\t%AN_fin\\t%AF\\t%AF_afr\\t%AF_amr\\t%AF_nfe\\t%AF_asj\\t%AF_sas\n",
    "\\t%AF_eas\\t%AF_mid\\t%AF_fin\\t%vep=\\n' filtered_gnomad22.bgz > processed_gnomad22.vcf.bgz\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "09f204f7-7119-43ad-95f8-6a50766edd32",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import cyvcf2\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9467dd40-a582-45a9-9a67-ae276add6aec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted data has been saved to gnomad_data.csv\n"
     ]
    }
   ],
   "source": [
    "# Input VCF.bgz file path\n",
    "vcf_path = 'data/test.bgz'\n",
    "output_csv_path = 'gnomad_data.csv'\n",
    "\n",
    "# Define columns to extract\n",
    "columns_to_extract = ['CHROM', 'POS', 'ID', 'REF', 'ALT']\n",
    "\n",
    "# Define INFO fields to extract\n",
    "info_fields_to_extract = ['AC', 'AC_afr', 'AC_amr', 'AC_nfe', 'AC_asj', 'AC_sas',\n",
    "                         'AC_eas', 'AC_mid', 'AC_fin', 'AN', 'AN_afr', 'AN_amr', 'AN_nfe', 'AN_asj', 'AN_sas',\n",
    "                         'AN_eas', 'AN_mid', 'AN_fin', 'AF', 'AF_afr', 'AF_amr', 'AF_nfe',\n",
    "                         'AF_asj', 'AF_sas', 'AF_eas', 'AF_mid', 'AF_fin']\n",
    "\n",
    "# Open VCF.bgz file using cyvcf2\n",
    "vcf_reader = cyvcf2.VCF(vcf_path)\n",
    "\n",
    "# Extract data from the VCF.bgz file\n",
    "data = []\n",
    "for variant in vcf_reader:\n",
    "    variant_data = [variant.CHROM, variant.POS, variant.ID, variant.REF, variant.ALT[0]]\n",
    "    info_data = [variant.INFO.get(field, '.') for field in info_fields_to_extract]\n",
    "    vep_annotation = variant.INFO.get('vep')\n",
    "    if vep_annotation:\n",
    "        vep_info_combined = []\n",
    "        for annotation in vep_annotation.split(','):\n",
    "            vep_info = annotation.split('|')\n",
    "            vep_info_combined.append(','.join([vep_info[i] if i < len(vep_info) else '' for i in [3, 1, 5, 7, 8, 9, 42, 43, 44]]))\n",
    "        data.append(variant_data + info_data + [';'.join(vep_info_combined)])\n",
    "    else:\n",
    "        data.append(variant_data + info_data + [''])  # Adding empty values if VEP annotation is not present\n",
    "\n",
    "# Save the extracted data into a CSV file\n",
    "with open(output_csv_path, 'w', newline='') as csv_file:\n",
    "    writer = csv.writer(csv_file)\n",
    "    header = columns_to_extract + info_fields_to_extract + ['VEP_annotation']\n",
    "    writer.writerow(header)\n",
    "    writer.writerows(data)\n",
    "\n",
    "print(f\"Extracted data has been saved to {output_csv_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a33cfa88-b752-41e0-aa93-0939e303b902",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'FRG1FP,non_coding_transcript_exon_variant,Transcript,unprocessed_pseudogene,9/9,,,,;FRG1FP,downstream_gene_variant,Transcript,transcribed_pseudogene,,,,,'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "var_data = pd.read_csv('gnomad_data.csv', sep = ',')  \n",
    "display(var_data.iloc[30]['VEP_annotation'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8c1c07e2-a7a1-4eeb-9f09-744856059235",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AC_afr</th>\n",
       "      <th>AC_amr</th>\n",
       "      <th>AC_nfe</th>\n",
       "      <th>AC_asj</th>\n",
       "      <th>...</th>\n",
       "      <th>AF_fin</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Feature_Type</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>EXON</th>\n",
       "      <th>INTRON</th>\n",
       "      <th>LoF</th>\n",
       "      <th>LoF_filter</th>\n",
       "      <th>LoF_flags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr22</td>\n",
       "      <td>10736362</td>\n",
       "      <td>rs879131185</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.20000000298023224</td>\n",
       "      <td>U2</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>snRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr22</td>\n",
       "      <td>10736399</td>\n",
       "      <td>rs1206833908</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>442</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>298</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5</td>\n",
       "      <td>U2</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>snRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr22</td>\n",
       "      <td>10736400</td>\n",
       "      <td>rs1289580688</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.022727299481630325</td>\n",
       "      <td>U2</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>snRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr22</td>\n",
       "      <td>10736414</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>U2</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>snRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22</td>\n",
       "      <td>10736415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>U2</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>snRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2367</th>\n",
       "      <td>chr22</td>\n",
       "      <td>15562681</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1/1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2367</th>\n",
       "      <td>chr22</td>\n",
       "      <td>15562681</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2368</th>\n",
       "      <td>chr22</td>\n",
       "      <td>15562681</td>\n",
       "      <td>rs1986857581</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.9197499568690546e-05</td>\n",
       "      <td></td>\n",
       "      <td>intron_variant&amp;non_coding_transcript_variant</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td></td>\n",
       "      <td>2/3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2368</th>\n",
       "      <td>chr22</td>\n",
       "      <td>15562681</td>\n",
       "      <td>rs1986857581</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.9197499568690546e-05</td>\n",
       "      <td></td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1/1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2368</th>\n",
       "      <td>chr22</td>\n",
       "      <td>15562681</td>\n",
       "      <td>rs1986857581</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.9197499568690546e-05</td>\n",
       "      <td></td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4881 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM       POS            ID REF ALT   AC  AC_afr  AC_amr  AC_nfe  \\\n",
       "0     chr22  10736362   rs879131185   C   T   63       0       1      50   \n",
       "1     chr22  10736399  rs1206833908   T   C  442      13       5     298   \n",
       "2     chr22  10736400  rs1289580688   G   A   15       0       0       9   \n",
       "3     chr22  10736414           NaN   G   C    1       0       0       1   \n",
       "4     chr22  10736415           NaN   C   T    1       0       0       1   \n",
       "...     ...       ...           ...  ..  ..  ...     ...     ...     ...   \n",
       "2367  chr22  15562681           NaN   C   A    1       0       0       0   \n",
       "2367  chr22  15562681           NaN   C   A    1       0       0       0   \n",
       "2368  chr22  15562681  rs1986857581   C   T    4       0       0       2   \n",
       "2368  chr22  15562681  rs1986857581   C   T    4       0       0       2   \n",
       "2368  chr22  15562681  rs1986857581   C   T    4       0       0       2   \n",
       "\n",
       "      AC_asj  ...                  AF_fin  SYMBOL  \\\n",
       "0          0  ...     0.20000000298023224      U2   \n",
       "1          5  ...                     0.5      U2   \n",
       "2          0  ...    0.022727299481630325      U2   \n",
       "3          0  ...                     0.0      U2   \n",
       "4          0  ...                     0.0      U2   \n",
       "...      ...  ...                     ...     ...   \n",
       "2367       0  ...                     0.0           \n",
       "2367       0  ...                     0.0           \n",
       "2368       0  ...  1.9197499568690546e-05           \n",
       "2368       0  ...  1.9197499568690546e-05           \n",
       "2368       0  ...  1.9197499568690546e-05           \n",
       "\n",
       "                                       Consequence  Feature_Type  \\\n",
       "0                            upstream_gene_variant    Transcript   \n",
       "1                            upstream_gene_variant    Transcript   \n",
       "2                            upstream_gene_variant    Transcript   \n",
       "3                            upstream_gene_variant    Transcript   \n",
       "4                            upstream_gene_variant    Transcript   \n",
       "...                                            ...           ...   \n",
       "2367            non_coding_transcript_exon_variant    Transcript   \n",
       "2367                         upstream_gene_variant    Transcript   \n",
       "2368  intron_variant&non_coding_transcript_variant    Transcript   \n",
       "2368            non_coding_transcript_exon_variant    Transcript   \n",
       "2368                         upstream_gene_variant    Transcript   \n",
       "\n",
       "                     BIOTYPE  EXON  INTRON  LoF  LoF_filter  LoF_flags  \n",
       "0                      snRNA                                            \n",
       "1                      snRNA                                            \n",
       "2                      snRNA                                            \n",
       "3                      snRNA                                            \n",
       "4                      snRNA                                            \n",
       "...                      ...   ...     ...  ...         ...        ...  \n",
       "2367    processed_pseudogene   1/1                                      \n",
       "2367                  lncRNA                                            \n",
       "2368  unprocessed_pseudogene           2/3                              \n",
       "2368    processed_pseudogene   1/1                                      \n",
       "2368                  lncRNA                                            \n",
       "\n",
       "[4881 rows x 41 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "var_data_proc = var_data['VEP_annotation'].str.split(';', expand=True).stack().reset_index(level=1, drop=True).to_frame('VEP')\n",
    "result = var_data.drop('VEP_annotation', axis=1).join(var_data_proc)\n",
    "\n",
    "new_columns = result['VEP'].str.split(',', expand=True)\n",
    "result[new_columns.columns] = new_columns\n",
    "\n",
    "result = result.drop('VEP', axis=1)\n",
    "\n",
    "\n",
    "result = result.rename(columns={0: 'SYMBOL', \n",
    "                                1: 'Consequence', \n",
    "                                2: 'Feature_Type', \n",
    "                                3: 'BIOTYPE', \n",
    "                                4: 'EXON', \n",
    "                                5: 'INTRON', \n",
    "                                6: 'LoF', \n",
    "                                7: 'LoF_filter',\n",
    "                                8: 'LoF_flags'})\n",
    "\n",
    "display(result)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BI_main_project",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
