{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5b2f6175-7143-4c75-89c1-5bdd1d0e4e5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import gzip\n",
    "import os\n",
    "import sys\n",
    "\n",
    "# csv.field_size_limit(sys.maxsize)\n",
    "\n",
    "clinvar_columns = ['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO']\n",
    "clinvar_columns_dict = {col: idx for idx, col in enumerate(clinvar_columns)}\n",
    "\n",
    "clinvar_info_names = [\n",
    "    'AF_ESP', 'AF_EXAC', 'AF_TGP', 'ALLELEID', 'CLNDN', 'CLNDNINCL', 'CLNDISDB', 'CLNDISDBINCL',\n",
    "    'CLNHGVS', 'CLNREVSTAT', 'CLNSIG', 'CLNSIGCONF', 'CLNSIGINCL', 'CLNVC', 'CLNVCSO', 'CLNVI',\n",
    "    'DBVARID', 'GENEINFO', 'MC', 'ONCDN', 'ONCDNINCL', 'ONCDISDB', 'ONCDISDBINCL', 'ONC', 'ONCINCL',\n",
    "    'ONCREVSTAT', 'ONCCONF', 'ORIGIN', 'RS', 'SCIDN', 'SCIDNINCL', 'SCIDISDB', 'SCIDISDBINCL',\n",
    "    'SCIREVSTAT', 'SCI', 'SCIINCL'\n",
    "]\n",
    "clinvar_info_dict = {info: idx for idx, info in enumerate(clinvar_info_names)}\n",
    "\n",
    "vep_names = ('Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON|INTRON|'\n",
    "             'HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|'\n",
    "             'Existing_variation|DISTANCE|STRAND|FLAGS|SYMBOL_SOURCE|HGNC_ID|CANONICAL').split('|')\n",
    "vep_names_dict = {name: idx for idx, name in enumerate(vep_names)}\n",
    "\n",
    "headers = ['CHROM', 'POS', 'ID', 'REF', 'ALT', 'CLNSIG', 'CLNVC',  'GENEINFO', 'MC', \n",
    "           'Consequence', 'SYMBOL', 'Gene', 'Feature_type', 'Feature', 'BIOTYPE', 'cDNA_position', 'CANONICAL']\n",
    "\n",
    "\n",
    "def parse_clinvar_vcf(vcf_file, output_dir='', output_filename=''):\n",
    "    # Если папка не указана, сохраняем в текущую директорию\n",
    "    if output_dir == '':\n",
    "        output_dir = os.getcwd()\n",
    "    \n",
    "    # Если имя выходного файла не задано, используем имя входного файла, но с расширением .tsv\n",
    "    # Через basename получаем имя файла, а через splitext оставляем только название без расширения\n",
    "    if output_filename == '':\n",
    "        output_filename = os.path.splitext(os.path.basename(vcf_file))[0] + '.tsv'\n",
    "    else:\n",
    "        # Если не указано расширение файла (указано неверно), добавляем .tsv\n",
    "        if not output_filename.endswith('.tsv'):\n",
    "            output_filename += '.tsv'\n",
    "    \n",
    "    # Полный путь к выходному файлу\n",
    "    output_file = os.path.join(output_dir, output_filename)\n",
    "    \n",
    "    # Запись шапки таблицы в выходной файл\n",
    "    with open(output_file, 'w', newline='', encoding='utf-8') as table_file:\n",
    "        writer = csv.writer(table_file, delimiter='\\t')\n",
    "        writer.writerow(headers)\n",
    "\n",
    "    \n",
    "    with gzip.open(vcf_file, 'rt') as input_file, open(output_file, 'a', newline='', encoding='utf-8') as output_file:\n",
    "        vcf_reader = csv.reader(input_file, delimiter='\\t')\n",
    "        tsv_writer = csv.writer(output_file, delimiter='\\t')\n",
    "        \n",
    "        for line in vcf_reader:\n",
    "            parsed_data = parse_vcf_line(line)\n",
    "            for data in parsed_data:\n",
    "                tsv_writer.writerow(data)\n",
    "\n",
    "\n",
    "def parse_vcf_line(line):\n",
    "    parsed_data = []\n",
    "\n",
    "    if not line[clinvar_columns_dict['CHROM']].startswith('#'):\n",
    "        chrom = 'chr' + line[clinvar_columns_dict['CHROM']]\n",
    "        position = line[clinvar_columns_dict['POS']]\n",
    "        variation_id = line[clinvar_columns_dict['ID']]\n",
    "        ref = line[clinvar_columns_dict['REF']]\n",
    "        alt = line[clinvar_columns_dict['ALT']]\n",
    "\n",
    "        info = line[clinvar_columns_dict['INFO']].split(';')\n",
    "        for element in info:\n",
    "            for category in clinvar_info_dict:\n",
    "                if element.startswith(f'{category}='):\n",
    "                    value = element.split('=')\n",
    "                    clinvar_info_dict[category] = value[-1] \n",
    "        \n",
    "        vep_info = info[-1].split(',')\n",
    "        \n",
    "        for element in vep_info:\n",
    "            variant_info = element.split('|')\n",
    "\n",
    "            info_consequence = variant_info[vep_names_dict.get('Consequence')]\n",
    "            info_symbol = variant_info[vep_names_dict.get('SYMBOL')]\n",
    "            info_gene = variant_info[vep_names_dict.get('Gene')]\n",
    "            info_feature_type = variant_info[vep_names_dict.get('Feature_type')]\n",
    "            info_feature = variant_info[vep_names_dict.get('Feature')]\n",
    "            info_biotype = variant_info[vep_names_dict.get('BIOTYPE')]\n",
    "            info_cDNA = variant_info[vep_names_dict.get('cDNA_position')]\n",
    "            info_canonical = variant_info[vep_names_dict.get('CANONICAL')]\n",
    "            \n",
    "            filtered_data = [\n",
    "                chrom, position, variation_id, ref, alt,\n",
    "                clinvar_info_dict['CLNSIG'], clinvar_info_dict['CLNVC'], clinvar_info_dict['GENEINFO'], clinvar_info_dict['MC'],\n",
    "                info_consequence, info_symbol, info_gene, info_feature_type, \n",
    "                info_feature, info_biotype,\n",
    "                info_cDNA, info_canonical\n",
    "            ]\n",
    "            parsed_data.append(filtered_data)\n",
    "\n",
    "    return parsed_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9e504c3-e8bb-406a-a77b-5a516ce93a47",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e2a3682a-e66b-4427-9f01-9e382f8c6ba5",
   "metadata": {},
   "source": [
    "Указываем input файл (или путь к нему), а также имя output (необязательно, задаётся по умолчанию по названию вцф):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f2a50335-6a82-4d39-a172-ae6e3026cd2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "parse_clinvar_vcf('data_dir/clinvar/clinvar_20240331_anno.vcf.gz', output_filename='clinvar_20240331_anno.vcf.TEST')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bi_kernel",
   "language": "python",
   "name": "bi_kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
