{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fc565154-10df-46fa-a441-bb2e1e42e3a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO\n",
    "# Прописать путь сохранения аутпута, если указано название нового файла (по дефотлу сохраняется в папку с парсером)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "85c8432a-5aee-488e-aaf7-d6261a878f78",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import gzip\n",
    "import os\n",
    "\n",
    "# Column names for VCF file\n",
    "vcf_columns = ['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO']\n",
    "vcf_columns_dict = {col: idx for idx, col in enumerate(vcf_columns)}\n",
    "\n",
    "# Column names for VEP data\n",
    "vep_names = [\n",
    "    'Allele', 'Consequence', 'IMPACT', 'SYMBOL', 'Gene', 'Feature_type', 'Feature', 'BIOTYPE',\n",
    "    'EXON', 'INTRON', 'HGVSc', 'HGVSp', 'cDNA_position', 'CDS_position', 'Protein_position',\n",
    "    'Amino_acids', 'Codons', 'ALLELE_NUM', 'DISTANCE', 'STRAND', 'FLAGS', 'VARIANT_CLASS',\n",
    "    'SYMBOL_SOURCE', 'HGNC_ID', 'CANONICAL', 'MANE_SELECT', 'MANE_PLUS_CLINICAL', 'TSL', 'APPRIS',\n",
    "    'CCDS', 'ENSP', 'UNIPROT_ISOFORM', 'SOURCE', '', '', 'DOMAINS', 'miRNA', 'HGVS_OFFSET',\n",
    "    'PUBMED', 'MOTIF_NAME', 'MOTIF_POS', 'HIGH_INF_POS', 'MOTIF_SCORE_CHANGE', 'TRANSCRIPTION_FACTORS',\n",
    "    'LoF', 'LoF_filter', 'LoF_flags', 'LoF_info'\n",
    "]\n",
    "\n",
    "# Column names for population data\n",
    "population_names = (\n",
    "    'AC', 'AC_afr', 'AC_amr', 'AC_nfe', 'AC_asj', 'AC_sas', 'AC_eas', 'AC_mid', 'AC_fin',\n",
    "    'AN', 'AN_afr', 'AN_amr', 'AN_nfe', 'AN_asj', 'AN_sas', 'AN_eas', 'AN_mid', 'AN_fin',\n",
    "    'AF', 'AF_afr', 'AF_amr', 'AF_nfe', 'AF_asj', 'AF_sas', 'AF_eas', 'AF_mid', 'AF_fin'\n",
    ")\n",
    "population_dict = {pop: idx for idx, pop in enumerate(population_names)}\n",
    "\n",
    "# Headers for the output TSV file\n",
    "headers = [\n",
    "    'Chr', 'Position', 'rsID', 'Ref', 'Alt', 'AC', 'Impact', 'Consequence',\n",
    "    'Gene_symbol', 'Canonical_transcript', 'cDNA_position', 'LoF', 'LoF_flag', 'LoF_filter'\n",
    "]\n",
    "\n",
    "\n",
    "def parse_vcf(vcf_file, output_dir='', output_filename=''):\n",
    "    if not output_dir:\n",
    "        output_dir = os.getcwd()\n",
    "\n",
    "    if output_filename == '':\n",
    "        output_filename = os.path.splitext(os.path.basename(vcf_file))[0] + '.tsv'\n",
    "    else:\n",
    "        if not output_filename.endswith('.tsv'):\n",
    "            output_filename += '.tsv'\n",
    "\n",
    "    output_file = os.path.join(output_dir, output_filename)\n",
    "\n",
    "    with open(output_file, 'w', newline='', encoding='utf-8') as table_file:\n",
    "        writer = csv.writer(table_file, delimiter='\\t')\n",
    "        writer.writerow(headers)\n",
    "\n",
    "    with gzip.open(vcf_file, 'rt') as input_file:\n",
    "        vcf_reader = csv.reader(input_file, delimiter='\\t')\n",
    "\n",
    "        for line in vcf_reader:\n",
    "            if line[vcf_columns_dict['CHROM']].startswith('chr') and line[vcf_columns_dict['FILTER']] == 'PASS':\n",
    "                filtered_data = process_line(line)\n",
    "                write_to_output(output_file, filtered_data)\n",
    "\n",
    "def process_line(line):\n",
    "    chrom = line[vcf_columns_dict['CHROM']]\n",
    "    position = line[vcf_columns_dict['POS']]\n",
    "    rs_id = line[vcf_columns_dict['ID']]\n",
    "    ref = line[vcf_columns_dict['REF']]\n",
    "    alt = line[vcf_columns_dict['ALT']]\n",
    "    population_dict.update(get_population_data(line))\n",
    "    transcript_info = get_canonical_info(line)\n",
    "\n",
    "    filtered_data = [\n",
    "        chrom,\n",
    "        position,\n",
    "        rs_id,\n",
    "        ref,\n",
    "        alt,\n",
    "        population_dict['AC'],\n",
    "        ', '.join(transcript_info['IMPACT']),\n",
    "        ', '.join(transcript_info['Consequence']),\n",
    "        ', '.join(transcript_info['SYMBOL']),\n",
    "        ', '.join(transcript_info['Feature']),\n",
    "        ', '.join(transcript_info['cDNA_position']),\n",
    "        ', '.join(transcript_info['LoF']),\n",
    "        ', '.join(transcript_info['LoF_flags']),\n",
    "        ', '.join(transcript_info['LoF_filter'])\n",
    "    ]\n",
    "\n",
    "    return filtered_data\n",
    "\n",
    "def get_population_data(line):\n",
    "    column_with_info = line[vcf_columns_dict['INFO']].split(';')\n",
    "    frequency = {}\n",
    "    for element in column_with_info:\n",
    "        for pop in population_dict:\n",
    "            if element.startswith(f'{pop}='):\n",
    "                frequency[pop] = element.split('=')[-1]\n",
    "    return frequency\n",
    "\n",
    "def get_canonical_info(line):\n",
    "    transcript_info = {field: [] for field in ['IMPACT', 'Consequence', 'SYMBOL', 'Feature', 'cDNA_position', 'LoF', 'LoF_flags', 'LoF_filter']}\n",
    "    info = line[vcf_columns_dict['INFO']].split(';')\n",
    "    vep_info = info[-1].split('|')\n",
    "    feature_index = vep_names.index('Feature')\n",
    "    canonical_index = vep_names.index('CANONICAL')\n",
    "    symbol_index = vep_names.index('SYMBOL')\n",
    "    step = 47\n",
    "\n",
    "    for i in range(feature_index, len(vep_info), step):\n",
    "        transcript = vep_info[i]\n",
    "        canonical = vep_info[i + canonical_index - feature_index]\n",
    "        symbol = vep_info[i + symbol_index - feature_index]\n",
    "\n",
    "        if symbol and transcript.startswith('ENST') and canonical:\n",
    "            for field in transcript_info:\n",
    "                field_index = vep_names.index(field)\n",
    "                info = vep_info[i + field_index - feature_index]\n",
    "                if info:\n",
    "                    if field == 'Feature' or field == 'CANONICAL':\n",
    "                        transcript_info[field].append(transcript)\n",
    "                    else:\n",
    "                        transcript_info[field].append(info)\n",
    "    return transcript_info\n",
    "\n",
    "def write_to_output(output_file, data):\n",
    "    with open(output_file, 'a', newline='', encoding='utf-8') as output_file:\n",
    "        tsv_writer = csv.writer(output_file, delimiter='\\t')\n",
    "        tsv_writer.writerow(data)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97d63206-cc79-48d0-a94f-82faa1a619be",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2a3682a-e66b-4427-9f01-9e382f8c6ba5",
   "metadata": {},
   "source": [
    "Указываем input файл (или путь к нему), а также имя output (необязательно, задаётся по умолчанию по названию вцф):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5e2dd832-492d-4c76-a69a-67beecdcb710",
   "metadata": {},
   "outputs": [],
   "source": [
    "parse_vcf('../raw_data/example_chr21.bgz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "283bc135-3953-4238-9627-530558346c57",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chromosome_files = [f'gnomad.exomes.v4.0.sites.chr{i}.vcf.bgz' for i in range(1, 23)]\n",
    "\n",
    "# for file_name in chromosome_files:\n",
    "#     file_path = os.path.join('../raw_data', file_name)\n",
    "#     get_parse_vcf(file_path, output_dir='data_dir')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aff62976-4a1e-466d-a892-61df8415acdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "parse_vcf('../raw_data/gnomad.exomes.v4.0.sites.chr21.vcf.bgz')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bi_kernel",
   "language": "python",
   "name": "bi_kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
